src/systems/filecoin_vm/runtime/runtime.go

```diff
@@ -1,81 +1,424 @@
+	}
+
+	// TODO: Update gasUsed
+	TODO()
+
+	return
+}
+
+func (rtOuter *Runtime_I) _sendInternal(to addr.Address, input InvocInput, ignoreErrors bool) msg.MessageReceipt {
+	rtOuter._checkRunning()
+	rtOuter._checkStateLock(false)
+
+	toActor := rtOuter._globalStatePending().GetActor(to).State()
+
+	toActorCode, err := loadActorCode(toActor.CodeCID())
+	if err != nil {
+		rtOuter._throwError(exitcode.SystemError(exitcode.ActorCodeNotFound))
+	}
+
+	var toActorMethodGasBound msg.GasAmount
+	TODO() // TODO: obtain from actor registry
+	rtOuter._deductGasRemaining(toActorMethodGasBound)
+	// TODO: gasUsed may be larger than toActorMethodGasBound if toActor itself makes sub-calls.
+	// To prevent this, we would need to calculate the gas bounds recursively.
+
+	rtOuter._transferFunds(rtOuter._actorAddress(), to, input.Value())
+	rtOuter._incrementCallSeqNum(to)
+
+	rtInner := Runtime_Make(
+		rtOuter._globalStatePending(),
+		to,
+		input.Value(),
+		rtOuter._gasRemaining(),
+	)
+
+	invocOutput, gasUsed := _invokeMethodInternal(
+		rtInner.Impl(),
+		toActorCode,
+		input.Method(),
+		input.Params(),
+	)
+
+	rtOuter._refundGasRemaining(toActorMethodGasBound)
+	rtOuter._deductGasRemaining(gasUsed)
+
+	if !ignoreErrors && invocOutput.ExitCode().IsError() {
+		rtOuter._throwError(exitcode.SystemError(exitcode.MethodSubcallError))
+	}
+
+	if invocOutput.ExitCode().AllowsStateUpdate() {
+		rtOuter._globalStatePending_ = rtInner._globalStatePending()
+	}
+
+	return msg.MessageReceipt_Make(invocOutput, gasUsed)
+}
+
+func (rt *Runtime_I) Send(to addr.Address, input InvocInput) msg.MessageReceipt {
+	return rt._sendInternal(to, input, false)
+}
+
+func (rt *Runtime_I) SendAllowingErrors(to addr.Address, input InvocInput) msg.MessageReceipt {
+	return rt._sendInternal(to, input, true)
+}
+
+type MessageQueue []MessageQueueItem
+type MessageQueueItem struct {
+	to           addr.Address
+	input        InvocInput
+	ignoreErrors bool
+}
+
+func MessageQueueItem_Make(to addr.Address, input InvocInput, ignoreErrors bool) MessageQueueItem {
+	return MessageQueueItem{
+		to:           to,
+		input:        input,
+		ignoreErrors: ignoreErrors,
+	}
+}
+
+func (rt *Runtime_I) _deferredSendInternal(to addr.Address, input InvocInput, ignoreErrors bool) {
+	rt._checkRunning()
+	rt._checkStateLock(false)
+	rt._messageQueue_ = append(rt._messageQueue(), MessageQueueItem_Make(to, input, ignoreErrors))
+}
+
+func (rt *Runtime_I) DeferredSend(to addr.Address, input InvocInput) Runtime_DeferredSend_FunRet {
+	rt._deferredSendInternal(to, input, false)
+	return &Runtime_DeferredSend_FunRet_I{}
+}
+
+func (rt *Runtime_I) DeferredSendAllowingErrors(to addr.Address, input InvocInput) Runtime_DeferredSendAllowingErrors_FunRet {
+	rt._deferredSendInternal(to, input, true)
+	return &Runtime_DeferredSendAllowingErrors_FunRet_I{}
+}
+
+func (rt *Runtime_I) ValueSupplied() actor.TokenAmount {
+	return rt._valueSupplied()
+}
+
+func (rt *Runtime_I) Randomness(offset uint64) Randomness {
```

Comment:
- Keep `Randomness(e block.ChainEpoch, offset uint64)`
- Actors need to be able to access the randomness of past epochs (up to some K lookback). this is critical for SPC, PoSt, sealing, and more. 
- maybe `e` must be `K <= e <= now()` (or panics/errors)

Reply:
Done.